### PR TITLE
RD-5591 Migrate Redux store to TypeScript - part 2 (app, contexts, formErrors, plugins, userApp)

### PR DESCRIPTION
**Pull request by qooban**
## Description
This PR extends solution introduced in https://github.com/cloudify-cosmo/cloudify-stage/pull/2225.

The following actions and reducers are covered by this PR (additional work in brackets):
1. `app` (moved `actions/appState.ts` to `actions/app.ts` and moved `actions/app.ts` to `actions/initialPageLoad.ts`)
2. `context` and `drilldownContext`
3. `formErrors`
4. `plugins` (aligned reducer name with the convention we follow)
5. `userApp` (merged `actions/userAppCommon.ts` actions file into `actions/userApp.ts`)

In addition to that the following fixes/refactors were applied:
1. Simplified tenants list in Redux store
2. Improved `ReduxThunkAction` type to allow action-types restrictions
3. Updated tests

In the next PR I will focus on template management and final solution polishing.

## Screenshots / Videos
N/A

## Checklist
- [x] My code follows the [UI Code Style](https://cloudifysource.atlassian.net/l/c/x8fq902p).
- [x] I followed the [UI Testing Guidelines](https://cloudifysource.atlassian.net/l/cp/udCVfvrx) and verified that all tests/checks have passed.
- [x] I verified if that change requires any update in the [documentation](https://cloudifysource.atlassian.net/l/c/4XKtPocy).
- [x] I added proper labels to this PR.

## Tests
1. [Build Status](https://jenkins.cloudify.co/view/UI%20Health%20View/job/Stage-UI-System-Test/3749/) (27.10 11:22 CEST) - FAILED (a lot of issues, should be fixed by https://github.com/cloudify-cosmo/cloudify-stage/pull/2233/commits/b325af3d15ad083e3ddeb2a3b04daf716e83bf5f and https://github.com/cloudify-cosmo/cloudify-stage/pull/2233/commits/a195a434942fd1c2f930c3bd7aeadcd9a46f947b)

2. [Build Status](https://jenkins.cloudify.co/view/UI%20Health%20View/job/Stage-UI-System-Test/3754/) (27.10 12:43 CEST) - PASSED (failures only on known issue - RD-5755)

## Documentation
N/A

### CODE REVIEW

File: app/actions/userApp.ts [hunk 1]
```diff
@@ -74,19 +90,21 @@ export function loadOrCreateUserAppData() {
...
     return (dispatch, getState) => {
         dispatch(setAppLoading(true));
 
         return dispatch(loadOrCreateUserAppData()).then(() => {
-            const getPageById = (pages, pageId) => {
-                return _.find(pages, { id: pageId });
+            const getPageById = (pages: PageMenuItem[], pageId?: string | null) => {
+                // TODO(RD-5591): I suspect a bug here. We should probably take into account Page Groups as well
+                return _.find(pages, { id: pageId }) as PageDefinition | undefined;
             };
```
Comment (kubama) [inline app/actions/userApp.ts:101]: Aside form the possible bug: `PageMenuItem` and `PageDefinition` types are unrelated in terms of types hierarchy. Seeing `PageMenuItem` being cast to `PageDefinition` I'm wondering - maybe these types could/should be part of the same hierarchy?

**Comment by Vorbert-Kruk** [comment]
Today I won't make it in time to do a review
I'll do it tomorrow morning 🎸

File: app/actions/app.ts [hunk 2]
```diff
@@ -1,43 +1,40 @@
...
 }
+
+export function setAppLoading(isLoading: boolean): SetAppLoadingAction {
+    return {
+        type: ActionType.SET_APP_LOADING,
+        payload: isLoading
+    };
+}
+
+export function setAppError(error: string | null) {
```
Comment (Vorbert-Kruk) [inline app/actions/app.ts:27]:
```suggestion
export function setAppError(error: string | null): SetAppErrorAction {
```

File: app/actions/manager/tenants.ts [hunk 3]
```diff
@@ -3,16 +3,15 @@ import type { Action } from 'redux';
...
 import Manager from '../../utils/Manager';
-import { setAppLoading } from '../appState';
+import { setAppLoading } from '../app';
 import { setEditMode } from '../config';
 import { clearContext } from '../context';
 import { reloadUserAppData } from '../userApp';
+import type { PaginatedResponse } from '../../../backend/types';
 
-// TODO(RD-5591): Refactor this to store { names: string[] } instead of { items: { name: string }[] }
-type Tenants = { items: { name: string }[] };
```
Comment (Vorbert-Kruk) [inline app/actions/manager/tenants.ts:12]: ### Optional
WDYT about still using `Tenants` type? 🤔 
 
Surely, we can freely use `array[]` in the functions below, but IMO having a dedicated type declaration for it (which would be used in multiple places) would make the code more maintanable and readable

File: app/actions/userApp.ts [hunk 4]
```diff
@@ -1,24 +1,38 @@
...
-        receivedAt: Date.now()
+        payload: {
+            pages,
+            receivedAt: Date.now()
+        }
     };
 }
 
-export function resetPages() {
+export function resetPages(): ThunkAction<void, ReduxState, never, AnyAction> {
```
Comment (Vorbert-Kruk) [inline app/actions/userApp.ts:35]: In #2225 [we've introduced](https://github.com/cloudify-cosmo/cloudify-stage/pull/2225/commits/a0560664de080380744930a16a8fab036d81d35a) `ReduxThunkAction` type
We could reuse it here 💪

File: app/reducers/managerReducer/tenantsReducer.ts [hunk 5]
```diff
@@ -1,14 +1,12 @@
 import type { Reducer } from 'redux';
-import _ from 'lodash';
```
Comment (Vorbert-Kruk) [inline app/reducers/managerReducer/tenantsReducer.ts:2]: Always good to see those kinds of changes 🥳

**Review by Vorbert-Kruk — COMMENTED** [review]
Nicely done migration!
Kudos for additionally doing `tenants` refactor, it's making the application logic fair simpler 🎉

File: app/actions/app.ts (hunk 2 above)
Comment (qooban) [reply app/actions/app.ts]: Done.

File: app/actions/userApp.ts (hunk 4 above)
Comment (qooban) [reply app/actions/userApp.ts]: Done.

File: app/actions/userApp.ts (hunk 1 above)
Comment (qooban) [reply app/actions/userApp.ts:104]: Right. In #2235 I'm focusing on all UI layouts (template management, page menu, pages, page groups and templates) related stuff, so let me postpone that change to the next PR.

File: app/actions/manager/tenants.ts (hunk 3 above)
Comment (qooban) [reply app/actions/manager/tenants.ts:12]: Sure, added.

File: app/actions/userApp.ts [hunk 6]
```diff
@@ -102,3 +123,25 @@ export function reloadUserAppData() {
         });
     };
 }
+
+export function saveUserAppData(): ReduxThunkAction<Promise<PostUserAppResponse>, never> {
```
Comment (qooban) [inline app/actions/userApp.ts:127]: Moved it here from `userAppCommon.ts` file.

**Review by qooban — COMMENTED** [review]
Answered ongoing discussions and since your last review I also improved `ReduxThunkAction` type.

PR is ready for final reviews.

**Review by Vorbert-Kruk — APPROVED** [review]
**Ḽ̴̛̟̙̽̇̄̒̀̈̿͘̚͠ǫ̵̛̠̙͓̮̳̰̳̭̭̣̗͎͉̗̎͌̉̉̈́̓̀͂͂͂̀ͅo̵̞̳̳̥̹͔͔͖͖̭̫̥̤͋̕k̵̡̗̎̀͌̌̂̒̉͊̕͠͠͠͝͠͝s̷̰̍̾̾͐̒͂̓̈́͋̅͂̑̈́̃͂͛̕͘̕͝͝ ̴̟̂̓̀̂͐̃̎͠g̵̛͎̞͓͚̣̲̝̩̭̤̭͖̝͔̓̄̽̇̉̏͛̃̎́̑̌͛͊̒̄̕͝r̷̡̩͔̯̥̫̳̰̟̳͙̤͉̲͇̩̂́̅̋̈́͐̎̉͒̊̓̎̌̕͝ȩ̷̨̢̮̩̗̭̖̰̬͖̲̽̈́̾̔̌̒̅̓͊̈̊̍͂ą̶̛̹͓̬͖̗͎̟͚̭͕̘̪̼͚͖͛̃͑̋͂̉̇̋̐ͅt̸̖̝͇̮͈̘͗ͅ** 🎉 
Thanks for addressing all comments 🎸

**Review by kubama — APPROVED** [review]

**Pull request merged**